src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                                              ImmutableDictionary<string, string>>;
+
+    /// <summary>
+    /// Analyzer to report unused expression values and parameters:
+    /// It flags the following cases:
+    ///     1. Expression statements that drop computed value, for example, "Computation();".
+    ///        These should either be removed (redundant computation) or should be replaced
+    ///        with explicit assignment to discard variable OR an unused local variable,
+    ///        i.e. "_ = Computation();" or "var unused = Computation();"
+    ///     2. Value assignments to locals/parameters that are never used on any control flow path,
+    ///        For example, value assigned to 'x' in first statement below is unused and will be flagged:
+    ///             x = Computation();
+    ///             if (...)
+    ///                 x = Computation2();
+    ///             else
+    ///                 Computation3(out x);
+    ///             ... = x;
+    ///        Just as for case 1., these should either be removed (redundant computation) or
+    ///        should be replaced with explicit assignment to discard variable OR an unused local variable,
+    ///        i.e. "_ = Computation();" or "var unused = Computation();"
+    ///     3. Redundant parameters that fall into one of the following two categories:
+    ///         a. Have no references in the executable code block(s) for it's containing method symbol.
+    ///         b. Have one or more references but it's initial value at start of code block is never used.
+    ///            For example, if 'x' in the example for case 2. above was a parameter symbol with RefKind.None
+    ///            and "x = Computation();" is the first statement in the method body, then it's initial value
+    ///            is never used. Such a parameter should be removed and 'x' should be converted into a local.
+    ///        We provide additional information in the diagnostic message to clarify the above two categories
+    ///        and also detect and mention about potential breaking change if the containing method is a public API.
+    ///        Currently, we do not provide any code fix for removing unused parameters as it needs fixing the
+    ///        call sites and any automated fix can lead to subtle overload resolution differences,
+    ///        though this may change in future.
+    /// </summary>
+    internal abstract class AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private const string UnusedValuePreferenceKey = nameof(UnusedValuePreferenceKey);
+        private const string IsUnusedLocalAssignmentKey = nameof(IsUnusedLocalAssignmentKey);
+        private const string IsRemovableAssignmentKey = nameof(IsRemovableAssignmentKey);
+
+        // "Expression value is never used"
```

Comment:
this comment just seems to be restating the loc string.  if this was to be commented it would likely be good to explain which case this is covering.